x/il/keeper/abci.go

```diff
@@ -0,0 +1,129 @@
+package keeper
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/armon/go-metrics"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	peggytypes "github.com/althea-net/peggy/module/x/peggy/types"
+
+	"github.com/peggyjv/sommelier/x/il/types"
+	oracletypes "github.com/peggyjv/sommelier/x/oracle/types"
+)
+
+// EndBlocker is called at the end of every block
+func (k Keeper) EndBlocker(ctx sdk.Context) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
+
+	params := k.GetParams(ctx)
+
+	// TODO: this should fetch an oracle data per identifier
+	oracleData := k.oracleKeeper.GetOracleData(ctx, oracletypes.UniswapDataType)
+	if oracleData == nil {
+		return
+	}
+
+	uniswapData, ok := oracleData.(*oracletypes.UniswapData)
+	if !ok {
+		return
+	}
+
+	pairs, err := uniswapData.Parse()
+	if err != nil {
+		// this shouldn't happen
+		panic(err)
+	}
+
+	pairMap := make(map[string]oracletypes.UniswapPairParsed)
+	for _, pair := range pairs {
+		pairMap[pair.ID] = pair
+	}
+
+	k.IterateStoplossPositions(ctx, func(address sdk.AccAddress, stoploss types.Stoploss) (stop bool) {
+		//   TODO: fetch pair info from oracle given the pair address
+		pair, ok := pairMap[stoploss.UniswapPairId]
+		if !ok {
+			// pair not found, continue with the next position
+			return false
+		}
+
+		reserveStr0 := strconv.FormatFloat(pair.Reserve0, 'f', 2, 64)
+		if reserveStr0 == "" {
+			return false
+		}
+
+		reserveStr1 := strconv.FormatFloat(pair.Reserve1, 'f', 2, 64)
+		if reserveStr1 == "" {
+			return false
+		}
+
+		reserve0, err := sdk.NewDecFromStr(reserveStr0)
+		if err != nil {
+			return false
+		}
+
+		reserve1, err := sdk.NewDecFromStr(reserveStr1)
+		if err != nil {
+			return false
+		}
+
+		currentRatio := reserve0.Quo(reserve1)
+
+		//   calculate current_ratio/reference_pair_ratio
+		currentSlippage := currentRatio.Quo(stoploss.ReferencePairRatio)
+
+		if currentSlippage.LTE(stoploss.MaxSlippage) {
+			// threshold not met, continue with the next position
+			return false
+		}
+
+		// Since the current slipage is grater than max slippage we now withdraw the liquidity of the LP for the uniswap pair
```

Comment:
I remember a few weeks back, we talked about different ways to model the stop-loss trigger:

* USD value - e.g., 'I put 10k into this pool, stop me out if the total position value falls below 9k'
* Token ratio - e.g. 'I put 100 UNI and 2000 DAI into this pool, stop me out if I fall below 90 UNI')

This looks like a calculation on the second bullet point - is the first bullet point covered by the `currentSlippage` calculation, or is that no longer a model of stop loss we plan to expose? 

I think we could do that with something like `usdValueOfPosition = (lpTokensOfUser/totallpTokensInPool) * reserveUSD`, but maybe that's not an interesting way of modeling LP exposure?

Reply:
we might be mixing the two here @kkennis. Will read through again tomorrow.